### PR TITLE
Implement cf login for CF-on-K8s

### PR DESCRIPTION
**Pull request by gcapizzi**
This implements a CF-on-K8s-compatible version of the `cf login` command. When a user runs `cf login` against a CF-on-K8s, they will be prompted to choose an `auth-info` from the ones specified in their `$KUBECONFIG`. Passing any extra option, except for `-a`, will result in a failure.

The main changes here are the extraction of the `AuthActor` and `UserConfig` interfaces, which we implement both for the default use case and the Kubernetes one. This allow us to provide specific implementations of methods like `Authenticate`, `GetLoginPrompts` and `CurrentUser*`.

While the choice of the `AuthActor` implementation can happen in the `NewActor` constructor, this can't be done for `UserConfig`. `cf login` hits the root endpoint and mutates the `Config` accordingly, and the selection of the `UserConfig` implementation depends on those mutations. We have then decided to perform this choice dynamically via the
`DynamicUserConfig` decorator.

### CODE REVIEW

File: actor/v7action/auth.go [hunk 1]
```diff
@@ -11,36 +11,48 @@ import (
 	"code.cloudfoundry.org/cli/cf/configuration/coreconfig"
 )
 
-func (actor Actor) Authenticate(credentials map[string]string, origin string, grantType constant.GrantType) error {
-	if grantType == constant.GrantTypePassword && actor.Config.UAAGrantType() == string(constant.GrantTypeClientCredentials) {
+type defaultAuthActor struct {
+	config    Config
+	uaaClient UAAClient
+}
```
[the remark this answers is not in the document]
Comment (jdgonzaleza) [reply actor/v7action/auth.go:17]: Never mind... I think they are not being used elsewhere, right?

File: go.mod [hunk 2]
```diff
@@ -10,64 +10,55 @@ require (
 	code.cloudfoundry.org/clock v1.0.0
 	code.cloudfoundry.org/diego-ssh v0.0.0-20170109142818-18cdb3586e7f
 	code.cloudfoundry.org/go-log-cache v1.0.1-0.20200316170138-f466e0302c34
-	code.cloudfoundry.org/go-loggregator v7.4.0+incompatible
+	code.cloudfoundry.org/go-loggregator/v8 v8.0.5
```
[the remark this answers is not in the document]
Comment (danail-branekov) [reply go.mod:13]: We had to use a reasonably new version of client-go to be able to use the kubernetes API. Unfortunately, client-go requires a newer version of protobuf which truned out not to be compatible with go-log-cache. In order to address this we had to [PR](https://github.com/cloudfoundry/go-log-cache/pull/31) go-log-cache and temporarily use the PR-ed version here (i.e. in the CLI). In the end, latest go-log-cache carries the dependency to `code.cloudfoundry.org/go-loggregator/v8 `

File: actor/v7action/auth.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdgonzaleza) [reply actor/v7action/auth.go:17]: That's what I was thinking.. I realized while looking at other files. I think this way is cleaner . Thanks for the confirmation!

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdgonzaleza) [reply go.mod:13]: Thanks for sharing the context!

File: actor/v7action/actor.go [hunk 3]
```diff
@@ -35,12 +44,18 @@ func NewActor(
...
 	clk clock.Clock,
 ) *Actor {
+	authActor := NewDefaultAuthActor(config, uaaClient)
+	if config != nil && config.IsCFOnK8s() {
+		authActor = NewKubernetesAuthActor(config, NewDefaultKubernetesConfigGetter())
+	}
+
 	return &Actor{
 		CloudControllerClient: client,
 		Config:                config,
```
Comment (jdgonzaleza) [inline actor/v7action/actor.go:54]: Do we still need the config and uaaClient fields in the Actor. It seems like we are now passing them into the AuthActor. with: @hjcalderon10

**Pull request merged**